### PR TITLE
Process.Linux: handle when /proc and the process pid namespace don't match.

### LINKED ISSUE #99887: Process.Linux: handle when proc mount mismatches with pid namespace
In scenarios with (nested) rootless containers, due to limited permissions, a container may have a new pid namespace but still have a parent's mount for `/proc`.

This causes pids under `/proc` to mismatch with those seen by the .NET process.

This issue is for making .NET work better in this specific scenario.

.NET can detect the scenario by reading the `/proc/self` symlink.
In the normal case (where `/proc` and the pid namespace match), the target path of the link is equal to the process pid.
When `/proc` is mounted for a parent, the target path will be the pid for the process in the parent pid namespace.

In both cases, .NET can read information about the current process by using `/proc/self` instead of `/proc/<pid>`.

In the second case, .NET shouldn't retrieve information about other processes from `/proc/<pid>` because it can't map between the `/proc` pid and the process pid namespace.

cc @omajid

### PR DESCRIPTION
**Pull request by tmds**
Normally the '/proc' filesystem uses the same pid namespace as the process. With rootless containers, it may happen that these pid namespaces do not match because the container doesn't have permissions to change '/proc' but it can create a new pid namespace.

When that happens, the numeric ids used by the /proc filesystem no longer match with the process pid namespace. We can still access information for the current process using '/proc/self'. For other processes, we can't map pids to the proc pids, which leads to reading information for non-existing/wrong/inaccessible processes.

Closes https://github.com/dotnet/runtime/issues/99887.

@dotnet/area-system-diagnostics-process @stephentoub ptal.

cc @omajid

### CODE REVIEW

**Comment by tmds** [comment]
> With rootless containers, it may happen that these pid namespaces do not match

To reproduces the issue without containers, you can run `sudo unshare -p --fork /bin/sh`.

In that environment you can run a program like:
```
Console.WriteLine(Process.GetCurrentProcess().PeakWorkingSet64);
```

When running this, sometimes this will print out a value, sometimes throw `InvalidOperationException` for an exited process.

In all cases, it is using information from `/proc` for another process than the current one.

**Comment by tmds** [comment]
@dotnet/area-system-diagnostics-process can you take a look?

Though it is not a common scenario, I think it's valuable we make .NET behave correct because it can occur in containers and CI. A user can not fix it and diagnosing the root cause may also be a challenge for them.

**Comment by am11** [comment]
For "self" case, we could short-circuit to `Environment.ProcessId` as we are doing that for `Environment.ProcessPath` in this code. (assuming `Environment.ProcessId` is returning the correct value in rootless case; I have't tested)

**Comment by tmds** [comment]
> For "self" case, we could short-circuit to Environment.ProcessId as we are doing that for Environment.ProcessPath in this code. (assuming Environment.ProcessId is returning the correct value in rootless case; I have't tested)

To get the current process pid we don't use `/proc`. I don't understand what this means.

**Comment by am11** [comment]
> I don't understand what you are suggesting.

I was looking at this and thought it could be replaced by  `procPid == Environment.ProcessId`:

image

**Comment by tmds** [comment]
> I was looking at this and thought it could be replaced by procPid == Environment.ProcessId:

We're adding `ProcPid` to represent identifiers as they are used by `/proc`. `Environment.ProcessId` is an identifier as used by the pid namespace. We mustn't compare these directly. `TryGetProcPid` gets us from `ProcessId` to `ProcPid.Self`. Then in `GetExePath` for `ProcPid.Self` instead of reading from `/proc/self` we "optimize" by using `Environment.ProcessPath`.

**Comment by tmds** [comment]
@dotnet/area-system-diagnostics-process ptal.

**Comment by tmds** [comment]
@dotnet/area-system-diagnostics-process ptal.

**Comment by tmds** [comment]
@jeffhandley @adamsitnik can you take a look at this PR please?

**Review by stephentoub — APPROVED** [review]
Sorry for the delay. LGTM.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs
Comment (adamsitnik) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs:283]: `Debug.Assert(procSelfPid.HasValue)` assumes that the value will be always present, but the condition below does not follow this assumption and in such case returns 1 ('/proc' and the process pid namespace match).

Why is that?

**Review by adamsitnik — COMMENTED** [review]
Overall LGTM, but please answer my question before I hit the merge button.

And once again big thanks for your help @tmds and apologies for the delays in my areas (I was 100% focused on the `BinaryFormatter` removal effort)

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs
Comment (tmds) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs:283]: It's handling the unexpected case where we were unable to determine the proc fs pid.

**Comment by tmds** [comment]
@adamsitnik this should be good to merge.

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @tmds !

**Comment by adamsitnik** [comment]
I need to close and re-open the PR so the CI can re-run the failed legs.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Linux.cs
Comment (am11) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Linux.cs:10]: This breaks the build on illumos. Was this intentional?

File: src/libraries/System.Private.CoreLib/src/System/Environment.SunOS.cs
Comment (am11) [inline src/libraries/System.Private.CoreLib/src/System/Environment.SunOS.cs:11]: This breaks the build on illumos. Was this intentional?

Comment (tmds) [reply src/libraries/System.Private.CoreLib/src/System/Environment.SunOS.cs:11]: No. I assume the fix is straight forward?

Comment (am11) [reply src/libraries/System.Private.CoreLib/src/System/Environment.SunOS.cs:11]: Yes, I will push it to my ongoing branch https://github.com/dotnet/runtime/pull/105178.

Comment (tmds) [reply src/libraries/System.Private.CoreLib/src/System/Environment.SunOS.cs:11]: Thanks!

**Pull request merged**
